### DISCUSSION TITLE
Meta set on server side gets lost on client side

### DISCUSSION THREAD

**Discussion by praveshishere** [nuxt/nuxt#16398] [Questions] [answered]
So, the scenario is like this

I fetch meta-information using an API request and then set it on the page using `useHead`, 
and I don't want to set it again on the client side since the page served from SSR already has all the meta information.

But it's not working, if I set meta only on the server side, once hydration kicks in, the whole head tag seems to reset.
I don't think meta or head tag is meant to be set twice or required to be set twice

Here is an example reproduction, where the set meta info on the server side is lost on the client side
https://stackblitz.com/edit/github-bnajer?file=app.vue&theme=dark

> Server Side
![image](https://user-images.githubusercontent.com/95841565/191905270-b48d6d3e-fd90-4052-82e7-cb6f507a7055.png)

> Client Side
![image](https://user-images.githubusercontent.com/95841565/191905361-62bf6eee-494f-4f3e-9091-4bcb47247fcd.png)

Is this an expected behavior, and if so, how can one disable it?

**Comment by TheAlexLichter** [comment]
If you wrap the call in `process.server`, then it will only be present on the server.
If you remove the condition, it shouldn't trigger another call on client-side and do exactly what you want ☺️

**Reply by praveshishere** [reply]
> If you wrap the call in `process.server`, then it will only be present on the server. If you remove the condition, it shouldn't trigger another call on the client side, and do exactly what you want

@manniL  removing the condition means, it would run on the client too (as I have debugged the code), I just want the head tag to stay intact even after hydration.

Also, I don't see any purpose in generally setting meta at both sides, once is sufficient.

**Reply by TheAlexLichter** [reply]
> Also, I don't see any purpose in generally setting meta at both sides, once is sufficient.

No, if you e.g. navigate from one site to another, site-specific meta tags have to be rendered on the client as well 😋

**Reply by TheAlexLichter** [reply]
Regarding hydration you are right - https://github.com/vueuse/head/issues/75 is the best-fitting open issue I'd say. If you think it doesn't fit, feel free to raise another one in the VueUse head repo 👍🏻

**Reply by praveshishere** [reply]
Thanks a lot, I should wait for them to fix it.

**Reply by jgupta** [reply]
https://github.com/vueuse/head/issues/75 is closed and it is not fixed with it. Have tested with RC12 edge. Created a new issue at https://github.com/vueuse/head/issues/140

**Comment by anhthang** [comment]
I'm having same issue, nothing inside head when I use `ssr: false`

```html
<head>
    <link rel="modulepreload" as="script" crossorigin href="[/_nuxt/@vite/client](view-source:http://localhost:3000/_nuxt/@vite/client)">
    <link rel="modulepreload" as="script" crossorigin href="[/_nuxt/Users/path-to-folder/node_modules/nuxt/dist/app/entry.mjs](view-source:http://localhost:3000/_nuxt/Users/path-to-folder/node_modules/nuxt/dist/app/entry.mjs)">
    <script>....</script>
</head>
```
I already tried all examples from Nuxt 3 getting started page (https://v3.nuxtjs.org/getting-started/seo-meta/) but still same. It's working well in my old repo using Nuxt 2

Below is my nuxt config
```json
{
    "ssr": false,
    "target": "static"
}
```

How I can fix this?

**Reply by TheAlexLichter** [reply]
With `ssr: false`, SEO tags will never be rendered in the head because that is part of SSR, which you disabled 😋

(Only possible exception: The defaults provided via `nuxt.config.js`, but haven't tested that)

PS: `target: static` is not used in Nuxt3 anymore. Use `nuxi generate` for SSG and remove the `ssr` option.

**Reply by anhthang** [reply]
> With `ssr: false`, SEO tags will never be rendered in the head because that is part of SSR, which you disabled 😋
Is there any way to add metatags? Because I remember I have some issues when I use `ssr: true` before but I will test again

> (Only possible exception: The defaults provided via `nuxt.config.js`, but haven't tested that)
I tested in this fille for both `head` and `meta` config but it's not working also

> PS: `target: static` is not used in Nuxt3 anymore. Use `nuxi generate` for SSG and remove the `ssr` option.
If that, I dont need these configuration, right? :D

Thanks for your support

**Reply by anhthang** [reply]
Hi @manniL 👋

It's working in development, but I'm getting errors like below when run build and preview command before deploy to production.

```sh
[nuxt] [request error] [unhandled] [500] Package subpath './helpers/esm/defineProperty.js' is not defined by "exports" in /Users/path/.output/server/node_modules/@babel/runtime/package.json imported from /Users/path/.output/server/chunks/app/server.mjs
  at new NodeError (node:internal/errors:393:5)
  at throwExportsNotFound (node:internal/modules/esm/resolve:340:9)
  at packageExportsResolve (node:internal/modules/esm/resolve:619:3)
  at packageResolve (node:internal/modules/esm/resolve:793:14)
  at moduleResolve (node:internal/modules/esm/resolve:859:20)
  at defaultResolve (node:internal/modules/esm/resolve:1074:11)
  at nextResolve (node:internal/modules/esm/loader:163:28)
  at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
  at ESMLoader.getModuleJob (node:internal/modules/esm/loader:425:18)
  at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
```

Do you know what's error and how to fix it?

**Reply by u007** [reply]
i got this error too when running pnpm run build and running the output server, but on pnpm run dev works fine

**Comment by harlan-zw** [comment, marked as answer]
Hey guys, thanks for the discussion. This is a planned feature for the v1 of @vueuse/head.

There isn't really a way to support this using the existing `useHead`, as client-side hydration will be required for anything dynamic and it's not possible to infer that type of input using that composable.

To implement, I'd be looking to add a new composable, something like `useHeadServer`, which would have the same input types. This would accompany logic to separate the tags which need to be hydrated when SSR.

This will also be useful for meta-defined in nuxt.config, as it will always be static it won't need hydration either, providing a performance improvement.

If you have any ideas about usage please include them in this discussion and I'll consider it when I start playing with it. (within the next 1-2 weeks)

**Reply by praveshishere** [reply]
HI @harlan-zw, I believe that would really help, just wanted to confirm something,

> This will also be useful for meta-defined in nuxt.config, as it will always be static it won't need hydration either, providing a performance improvement.

Would it be possible for the `useServerHead` composable to also set meta in a way, that doesn't require hydration same as nuxt.config, also will it also work in case I have meta set after `suspense:resolve`, like after an await call.

**Reply by harlan-zw** [reply]
Correct, that's the plan for it. 

Suspense should be fine afaik, if you can currently SSR the meta then it will be supported

**Comment by harlan-zw** [comment]
Hey everyone, just letting you know that v1 of @vueuse/head now resolves this discussion.

You can read about the update here: https://harlanzw.com/blog/vue-use-head-v1
